wcfsetup/install/files/lib/system/message/unfurl/UnfurlResponse.class.php

```diff
@@ -0,0 +1,294 @@
+        }
+
+        try {
+            $request = new Request('GET', $url, [
+                'range' => \sprintf('bytes=%d-%d', 0, self::MAX_SIZE - 1),
+            ]);
+            $response = self::getHttpClient()->send($request);
+
+            return new self($url, $response);
+        } catch (BadResponseException $e) {
+            $response = $e->getResponse();
+
+            switch ($response->getStatusCode()) {
+                case 400: // Bad Request
+                case 401: // Unauthorized
+                case 402: // Payment Required
+                case 403: // Forbidden
+                case 404: // Not Found
+                    $message = "Request failed with status code {$response->getStatusCode()}.";
+
+                    throw new UrlInaccessible($message, $response->getStatusCode(), $e);
+                    break;
+
+                default:
+                    throw new DownloadFailed("Could not download content.", $response->getStatusCode(), $e);
+            }
+        } catch (ClientExceptionInterface $e) {
+            throw new DownloadFailed("Could not download content.", 0, $e);
+        }
+    }
+
+    /**
+     * @throws ParsingFailed If the body cannot be parsed (e.g. the url is an image).
+     * @throws DownloadFailed If the url can not be downloaded. This can be a temporary error.
+     */
+    private function __construct(string $url, Response $response)
+    {
+        $this->url = $url;
+        $this->response = $response;
+
+        $this->readBody();
+        $this->readDomDocument();
+    }
+
+    /**
+     * Reads the body of the given url and converts the body to utf-8.
+     */
+    private function readBody(): void
+    {
+        $this->body = "";
+        while (!$this->response->getBody()->eof()) {
+            $this->body .= $this->response->getBody()->read(8192);
+
+            if ($this->response->getBody()->tell() >= self::MAX_SIZE) {
+                break;
+            }
+        }
+        $this->response->getBody()->close();
+
+        if (\mb_detect_encoding($this->body) !== 'UTF-8') {
+            try {
+                $this->body = StringUtil::convertEncoding(\mb_detect_encoding($this->body), 'UTF-8', $this->body);
+            } catch (Exception $e) {
+                throw new ParsingFailed(
+                    "Could not parse body, due an invalid charset. The Url could be an image.",
+                    0,
+                    $e
+                );
+            }
+        }
+    }
+
+    /**
+     * Creates the DomDocument.
+     *
+     * @throws ParsingFailed If the body cannot be parsed (e.g. the url is an JSON file).
+     */
+    private function readDomDocument(): void
+    {
+        \libxml_use_internal_errors(true);
+        $this->domDocument = new \DOMDocument();
+        if (!$this->domDocument->loadHTML('<?xml version="1.0" encoding="UTF-8"?>' . $this->body)) {
+            throw new ParsingFailed("Could not parse body.");
+        }
```

Comment:
Is there any documentation, what exactly an internal error is?

Reply:
The method name is a bit of a misnomer. `\libxml_use_internal_errors()` controls whether PHP errors should be generated or not. We don't want it to generate PHP errors, because that is super wonky.